File: product_of_array_except_self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> leftProduct(nums.size(), 0);

        int product = 1;
        for(int i=0; i < nums.size(); i++) {
            leftProduct[i] = product;
            product *= nums[i];
        }
        
        product = 1;
        for(int i=nums.size()-1; i >= 0; i--) {
            leftProduct[i] *= product;
            product *= nums[i];
        }
        return leftProduct;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)

File: product_of_array_except_self.go
func productExceptSelf(nums []int) []int {
    ans := make([]int, len(nums))

    product := 1
    for i := 0; i < len(nums); i++ {
        ans[i] = product
        product = product * nums[i]
    }

    product = 1
    for i := len(nums)-1; i >= 0; i-- {
        ans[i] = ans[i] * product
        product = product * nums[i]
    }
    return ans
}

// Time Complexity : O(n)
// Space Complexity : O(n)